src/main/java/jenkins/scm/api/mixin/ChangeRequestSCMHead3.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm.api.mixin;
+
+import javax.annotation.CheckForNull;
+
+/**
+ * Additional attributes of a {@link ChangeRequestSCMHead} that should have been in the original mixin but we are not
+ * targeting Java 8 so we cannot add the default methods to the interface and must have an ugly {@code 3} class instead.
+ * @since 2.5.0
+ */
+// TODO once Java 8 is baseline move method to ChangeRequestSCMHead with default return value,
+// TODO deprecate this interface and add @Restricted(NoExternalUse.class) (retain empty interface for binary compat)
+public interface ChangeRequestSCMHead3 extends ChangeRequestSCMHead2 {
+
+    /**
+     * Returns the commit identifier (e.g. sha1) of the source branch.
+     *
+     * @return the commit identifier of the source branch of this change request or @null if such a concept doesn't exists in the backing source control system.
+     */
+    @CheckForNull
+    String getSourceCommitId();
```

Comment:
The revision already contains this information. The head would change as the change request updates which is a antipattern

Reply:
How can I access the Revision from the `BranchNameContributor`? Is `projectFactory.getRevision(j)` the revision we want?

Reply:
That will return the revision the job last “discovered”

There may or may not be a build corresponding to that revision. It may be `null`

Reply:
But *why* should the name of the branch depend on the revision? That would change all the time. Users would object

Reply:
`BranchNameContributor` is badly named IMO. It is in fact contributing all pipeline env variables related to multi-branch builds (normal branches, tags, pull requests). Some of those variables are already changing between each build (like `CHANGE_AUTHOR`)

Reply:
There's nothing stopping your own plugin from having its own name contributor... and that would be faster too, no need to wait for branch-api... and you can have complete control over the environment variable names

Reply:
Still back to the point of how to get this properly:
* we could write our own `EnvironmentContributor`, but same as `BranchNameContributor`, what would be the correct way to get the SCMRevision that the about to be created build will refers to?
* we could instead put this logic in our [SimpleBuildWrapper](https://github.com/SonarSource/sonar-scanner-jenkins/blob/e0a26cecaf56671602cef60ddc84bf255283385a/src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java#L93), doing something like:
    
```
public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment)
    throws IOException, InterruptedException {
    SCMSource src = SCMSource.SourceByItem.findSource(build.getParent());
    SCMRevision revision = src != null ? SCMRevisionAction.getRevision(src, build) : null;
```

but then how to get the revision hash? I see that each implementation has its own way. For example in the [GitHub plugin](https://github.com/jenkinsci/github-branch-source-plugin/blob/e8469af7b665580dddfacc56825e023d6d9f43db/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java#L82) or [Bitbucket plugin](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/c8bbed32f1a7245c53753679f4c456667c774997/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java#L165)

We can't do that in our plugin without adding a dependency on each branch source impl (and I'm not sure their classes are intended to be stable).

Do you think it would be doable to expose this hash in the API, for example in ChangeRequestSCMRevision?

Reply:
`ChangeRequestSCMRevision.getTarget()` is the `SCMRevision` you are after for Git based source control you should be able to extract the revision from that as it should have a common Git baseclass...

Reply:
> ```
>     SCMSource src = SCMSource.SourceByItem.findSource(build.getParent());
>     SCMRevision revision = src != null ? SCMRevisionAction.getRevision(src, build) : null
> ```

Hmmm seems that there is no good reason why you need to look up the source to get the revision action... should be possible to get the SCMRevisionAction directly from the build.

Reply:
Build is actionable...

I think you can just do `build.getActions(SCMRevisionAction.class)` and then figure out which ones you are interested in...

(Hint: we suspect you still haven't noticed that a build can check out multiple repositories... or even multiple revisions of the same repository... and that might shake your "we need a single revision" world view)

Reply:
> ChangeRequestSCMRevision.getTarget() is the SCMRevision you are after

Isn't `ChangeRequestSCMRevision.getTarget()` the target of the PR? The javadoc of `ChangeRequestSCMHead::getTarget` says:
> Branch to which this change would be merged or applied if it were accepted

What I need is the source branch head commit revision.

> we suspect you still haven't noticed that a build can check out multiple repositories
I know, but similar to `GitHubBuildStatusNotification` and `BitbucketBuildStatusNotifications` we would like currently to focus on "simple" cases.

Reply:
So the revision itself should be that. I suspect what you really want is a Git-centric base interface of `ChangeRequestSCMRevision` in the Git plugin, e.g. `GitChangeRequestSCMRevision` and then get all the Git based SCM Sources to use that interface... either that or maintain special cases to extract from each of those plugin's custom types

Reply:
> Build is actionable...
> 
> I think you can just do `build.getActions(SCMRevisionAction.class)` and then figure out which ones you are interested in...
> 
> (Hint: we suspect you still haven't noticed that a build can check out multiple repositories... or even multiple revisions of the same repository... and that might shake your "we need a single revision" world view)

This doesn't seem to work at least in github-branch-source plugin, I added an EnvironmentContributor and that returns null

(all actions)
![image](https://user-images.githubusercontent.com/21194782/66383884-b403a180-e9b5-11e9-9eb4-c28cb34534d2.png)

Reply:
Ahhh... so pipeline stores it differently, but it is stored just the same. You can get access to it from Branch API through the parent's helpers IIRC

Reply:
I'll try have another dig around later on, I wasn't able to find it in anything I had directly available, but I'm sure there's away